### PR TITLE
GEODE-9289: Configuration compatible with pre-1.12.0 versions

### PR DESCRIPTION
**Pull request by nabarunnag**
* Configuration sent from 1.12.0 locators and later can be interpreted by pre 1.12.0 version locators

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

**Review by mhansonp — APPROVED** [review]
This looks good. I would really like to see a unit test of the serialization, just to verify that it does what it says. e.g. serial the thing, deserialize the thing, verify the values.

**Comment by nabarunnag** [comment]
@mhansonp I completely agree with you but in this scenario, the serialization and deserialization code is split between Apache Geode 1.10.0 and Apache Geode 1.12.0+ hence it will be not possible to write the unit test.

File: geode-core/src/main/java/org/apache/geode/management/internal/configuration/domain/Configuration.java [hunk 1]
```diff
@@ -183,20 +185,48 @@ public void fromData(DataInput in) throws IOException, ClassNotFoundException {
...
+      if (jarNames != null) {
+        // we are reading pre 1.12 data. So add each jar name to deployments
+        jarNames.stream()
+            .map(x -> new Deployment(x, null, null))
+            .forEach(deployment -> deployments.put(deployment.getDeploymentName(), deployment));
+      }
     }
   }
 
+  public void toDataPre_GEODE_1_12_0_0(DataOutput out) throws IOException {
```
Comment (DonalEvans) [inline geode-core/src/main/java/org/apache/geode/management/internal/configuration/domain/Configuration.java:204]: Would it be possible to first call this method in `toData()` and then add the two extra lines that have been added there, to keep the two methods consistent?

File: geode-core/src/main/java/org/apache/geode/management/internal/configuration/domain/Configuration.java [hunk 2]
```diff
@@ -183,20 +185,48 @@ public void fromData(DataInput in) throws IOException, ClassNotFoundException {
     propertiesFileName = DataSerializer.readString(in);
     gemfireProperties = DataSerializer.readProperties(in);
     HashSet<String> jarNames = DataSerializer.readHashSet(in);
-    if (jarNames != null) {
-      // we are reading pre 1.12 data. So add each jar name to deployments
-      jarNames.stream()
-          .map(x -> new Deployment(x, null, null))
-          .forEach(deployment -> deployments.put(deployment.getDeploymentName(), deployment));
-    } else {
+    try {
```
Comment (DonalEvans) [inline geode-core/src/main/java/org/apache/geode/management/internal/configuration/domain/Configuration.java:188]: Given that we now have a fromDataPre_X method, it should be unnecessary to have this try/catch block, because we can be sure that we'll be reading from a member with the correct version here.

File: geode-core/src/main/java/org/apache/geode/management/internal/configuration/domain/Configuration.java [hunk 3]
```diff
@@ -183,20 +185,48 @@ public void fromData(DataInput in) throws IOException, ClassNotFoundException {
...
+    DataSerializer.writeString(configName, out);
+    DataSerializer.writeString(cacheXmlFileName, out);
+    DataSerializer.writeString(cacheXmlContent, out);
+    DataSerializer.writeString(propertiesFileName, out);
+    DataSerializer.writeProperties(gemfireProperties, out);
+    HashSet<String> jarNames = new HashSet<>(deployments.keySet());
+    DataSerializer.writeHashSet(jarNames, out);
+  }
+
+  public void fromDataPre_GEODE_1_12_0_0(DataInput in) throws IOException, ClassNotFoundException {
```
Comment (DonalEvans) [inline geode-core/src/main/java/org/apache/geode/management/internal/configuration/domain/Configuration.java:214]: I know that the implementations differ when it comes to how to handle the `jarNames` set, but would it be possible to find a way to have `fromData()` first call this method and then do any additional work that's specific to the post-1.12.0 implementation, just to reduce code duplication?

**Comment by mhansonp** [comment]
> @mhansonp I completely agree with you but in this scenario, the serialization and deserialization code is split between Apache Geode 1.10.0 and Apache Geode 1.12.0+ hence it will be not possible to write the unit test.

I did this recently with Anil's help would it work? GatewaySenderAdvisorTest.java has an example. It may not work for you.. Either way, I still approved the change. It is not necessary, just nice, if the approach works at all...

File: geode-gfsh/src/upgradeTest/java/org/apache/geode/management/ConfigurationCompatibilityTest.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+  }
+
+  @Rule
+  public ClusterStartupRule clusterStartupRule = new ClusterStartupRule();
+
+  @Rule
+  public GfshCommandRule gfsh = new GfshCommandRule();
+
+  @Test
+  public void whenConfigurationIsExchangedBetweenMixedVersionLocatorsThenItShouldNotThrowExceptions()
```
Comment (DonalEvans) [inline geode-gfsh/src/upgradeTest/java/org/apache/geode/management/ConfigurationCompatibilityTest.java:58]: When I revert the changes in this PR, this test still passes, so I'm not sure that it's testing what you're intending it to test.

**Review by DonalEvans — CHANGES_REQUESTED** [review]

File: geode-core/src/main/java/org/apache/geode/management/internal/configuration/domain/Configuration.java (hunk 2 above)
Comment (nabarunnag) [reply geode-core/src/main/java/org/apache/geode/management/internal/configuration/domain/Configuration.java:188]: the catch part is when it is reading from a pre 1.12.0 version because it will hit the EOFException is trying to read the Configuration when it is being sent through EntryEventImpl from region (ClusterConfig region) replication.  EntryEventImpl does not send the version in the stream and hence the fromData_pre is not triggered.
But in the case of servers, they request it via function execution, version is sent through the stream and that's when the xxData_pre_XX is triggered.

File: geode-core/src/main/java/org/apache/geode/management/internal/configuration/domain/Configuration.java (hunk 1 above)
Comment (nabarunnag) [reply geode-core/src/main/java/org/apache/geode/management/internal/configuration/domain/Configuration.java:204]: I think I can completely remove them and let this work via the exception try catch model itself for the function execution too.,

File: geode-gfsh/src/upgradeTest/java/org/apache/geode/management/ConfigurationCompatibilityTest.java (hunk 4 above)
Comment (nabarunnag) [reply geode-gfsh/src/upgradeTest/java/org/apache/geode/management/ConfigurationCompatibilityTest.java:58]: I attempted this in the 1.12.0 branch : https://github.com/apache/geode/pull/6494
same code changes and if you revert it in there, the test fails complaining NPEs were encountered, let me check what changed in develop.

File: geode-core/src/main/java/org/apache/geode/management/internal/configuration/domain/Configuration.java [hunk 5]
```diff
@@ -183,20 +185,48 @@ public void fromData(DataInput in) throws IOException, ClassNotFoundException {
...
-          .map(x -> new Deployment(x, null, null))
-          .forEach(deployment -> deployments.put(deployment.getDeploymentName(), deployment));
-    } else {
+    try {
       // version of the data we are reading (1.12 or later)
       final Version version = Versioning.getVersion(VersioningIO.readOrdinal(in));
       if (version.isNotOlderThan(KnownVersion.GEODE_1_12_0)) {
         deployments.putAll(DataSerializer.readHashMap(in));
       }
+    } catch (EOFException ex) {
```
Comment (jinmeiliao) [inline geode-core/src/main/java/org/apache/geode/management/internal/configuration/domain/Configuration.java:194]: When the xxData_pre_XX is not triggered, the existence of jarNames should be enough to determine whether we are reading a pre1.12 data or not. Do we have to use a try/catch block here?

Comment (nabarunnag) [reply geode-core/src/main/java/org/apache/geode/management/internal/configuration/domain/Configuration.java:194]: Unfortunately, when EntryEventImpl is sent, the version is not sent in the stream, hence in that case pre_XX is not triggered even if it is sent from pre 1.12.0 version

File: geode-core/src/main/java/org/apache/geode/management/internal/configuration/domain/Configuration.java [hunk 6]
```diff
@@ -169,7 +170,8 @@ public void toData(DataOutput out) throws IOException {
     // Before 1.12, this code wrote a non-null HashSet of jarnames to the output stream.
     // As of 1.12, it writes a null HashSet to the stream, so that when we can still read the old
```
Comment (upthewaterspout) [inline geode-core/src/main/java/org/apache/geode/management/internal/configuration/domain/Configuration.java:171]: This comment seems outdated since you are now writing the hash set of jar names.

File: geode-gfsh/src/upgradeTest/java/org/apache/geode/management/ConfigurationCompatibilityTest.java (hunk 4 above)
Comment (nabarunnag) [reply geode-gfsh/src/upgradeTest/java/org/apache/geode/management/ConfigurationCompatibilityTest.java:58]: @DonalEvans @upthewaterspout https://github.com/apache/geode/pull/6526 is created to fix the DUnit runner to detect suspicious messages in logs of VMs when they are restarted.
There was a behavior change in 1.14 hence the tests was not failing in 1.14 but in 1.12

Comment (DonalEvans) [reply geode-gfsh/src/upgradeTest/java/org/apache/geode/management/ConfigurationCompatibilityTest.java:58]: Thanks for the update. Once that PR is merged, I'll re-check that the test fails without the changes then approve this PR.

**Review by DonalEvans — APPROVED** [review]
Confirmed that the fix in #6526 allows the test added in this PR to fail without the changes, so I'm approving this PR.

**Review by jinmeiliao — APPROVED** [review]

**Review by dschneider-pivotal — APPROVED** [review]
